storage/diemdb/src/event_store/mod.rs

```diff
@@ -238,6 +282,80 @@ impl EventStore {
             })
             .collect::<Result<Vec<_>>>()
     }
+
+    /// Finds the first event sequence number in a specified stream on which `comp` returns false.
+    /// (assuming the whole stream is partitioned by `comp`)
+    fn search_for_event_lower_bound<C>(
+        &self,
+        event_key: &EventKey,
+        mut comp: C,
+        ledger_version: Version,
+    ) -> Result<Option<u64>>
+    where
+        C: FnMut(&ContractEvent) -> Result<bool>,
+    {
+        let mut begin = 0u64;
+        let mut end = match self.get_latest_sequence_number(ledger_version, event_key)? {
+            Some(s) => s
+                .checked_add(1)
+                .ok_or_else(|| format_err!("event sequence number overflew."))?,
+            None => return Ok(None),
+        };
+
+        let mut count = end - begin;
```

Comment:
Why using a temp `counter` and not the common way of doing binary search where we check `while (begin <= end)`? I.e., as in https://dev-notes.eu/2020/03/Binary-Search-in-Rust/

Reply:
http://www.cplusplus.com/reference/algorithm/lower_bound/ 
https://github.com/microsoft/STL/blob/b52c3797e262cbce28166d92f9d1b8f2bfeb5ac5/stl/inc/algorithm#L6384-L6406
It's a very common implementation. 
And what you posted is an exact search, not lower_bound needed here.